### ISSUE TITLE
plot/plotters_backend: infinite loop when all samples are the same

### ISSUE THREAD

**Issue by elagergren-spideroak** [bheisler/criterion.rs#467]
This snippet below will cause an infinite loop when generating `pdf.svg`. Specifically, `plotters_backend::pdf` calls `Sample::new(&xs)`, which results in `xs_.min() == xs_.max()`. The variable `xs_` is passed to `set_secondary_coord`, which generates a `Cartesian2d`. On line 232, `MeshStyle.draw` is called, which then calls `target.draw_mesh`, which tries to create `x_label` and `y_label` by calling `self.draw_mesh_lines`. That, in turn, calls `self.drawing_area.draw_mesh`, which calls `self.coord.draw_mesh`. Here, `self.coord` is `Cartesian2d` (as generated by `set_secondary_coord`). Calling its `draw_mesh` method tries to create `xkp` and `ykp` by calling `self.logic_x.key_points` and `self.logic_y.key_points`, respectively. Those methods are generated by the `gen_key_points_cmp` macro, which never breaks the `'outer` label. The label is never broken because:

1. `scale` is zero because `range.1 == range.0`: `10 ** floor(log10(x, x)) == 0`
2. `next_left` is NaN because `scale/nxt == 0`: `rem_euclid(a, 0) = a-ceil(a/0)*0 = a-ceil(+inf)*0 = a-NaN = NaN`
3. Same with `next_right`.
4. And NaN is sticky, so `npoints` is NaN.
5. `NaN > max_points` is false.

Admittedly, it's contrived: it requires a custom benchmark that always returns the same number. But, that could be the result of a bug, which is how I ran into it. And it's not exactly straightforward to debug. And while I haven't looked into it further, ISTM that it's even reproducible if N isn't a constant.

I'm also not sure if it's a bug in Criterion or in `plotters`. I'm not sure if a zero coordinate is valid or not. 

```
use criterion::{
    criterion_group, criterion_main,
    measurement::{Measurement, ValueFormatter},
    Criterion, Throughput,
};
use std::time::Duration;

fn repro(c: &mut Criterion<MyMeasurement>) {
    c.bench_function("repro", |b| {
        b.iter_custom(|_| {
            return 42;
        })
    });
}

struct MyMeasurement;

impl Measurement for MyMeasurement {
    type Intermediate = u32;
    type Value = u32;

    fn start(&self) -> Self::Intermediate {
        return 0;
    }
    fn end(&self, i: Self::Intermediate) -> Self::Value {
        return i;
    }
    fn add(&self, v1: &Self::Value, v2: &Self::Value) -> Self::Value {
        return *v1 + *v2;
    }
    fn zero(&self) -> Self::Value {
        return 0;
    }
    fn to_f64(&self, value: &Self::Value) -> f64 {
        return *value as f64;
    }
    fn formatter(&self) -> &dyn ValueFormatter {
        return &MyMeasurementFormatter;
    }
}

struct MyMeasurementFormatter;

impl ValueFormatter for MyMeasurementFormatter {
    fn scale_values(&self, _: f64, _: &mut [f64]) -> &'static str {
        "x"
    }

    fn scale_throughputs(&self, _: f64, throughput: &Throughput, _: &mut [f64]) -> &'static str {
        match *throughput {
            _ => panic!("unsupported"),
        }
    }

    fn scale_for_machines(&self, _: &mut [f64]) -> &'static str {
        "x"
    }
}

fn alt_measurement() -> Criterion<MyMeasurement> {
    return Criterion::default()
        .with_measurement(MyMeasurement)
        .warm_up_time(Duration::new(0, 10))
        .without_plots();
}

criterion_group! {
    name = benches;
    config = alt_measurement();
    targets = repro,
}
criterion_main!(benches);
```

**Comment by bheisler** [comment]
Hey, thanks for the bug report! Well debugged.

I've been able to reproduce this, but it seems like the problem lies down in plotters; specifying a X/Y-range of `(42.0)..(42.0)` is weird, but I don't think it should break plotters. I could work around it by ensuring that there's always a non-zero width in the X/Y range I pass to plotters, but since it's plotters itself that generates those ranges I don't think I should have to do that.

What do you think, @38?

**Comment by mb** [comment]
I am also affected by this bug. After installing gnuplot the benchmark panics instead of an infinite loop:

```gdb
thread 'main' panicked at 'assertion failed: slice.len() > 1 && slice.iter().all(|x| !x.is_nan())', /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/criterion-0.3.4/src/stats/univariate/sample.rs:29:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:50:5
   3: criterion::plot::gnuplot_backend::pdf::pdf_small
   4: <criterion::plot::gnuplot_backend::Gnuplot as criterion::plot::Plotter>::pdf
   5: <criterion::html::Html as criterion::report::Report>::measurement_complete
   6: criterion::analysis::common
   7: criterion::benchmark_group::BenchmarkGroup<M>::bench_function
   8: uart::bench::benches
   9: uart::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by 38** [comment]
Hi there, sorry for the delay. The direct cause of this infinite loop is Plotters doesn't handle the zero-sized axis correctly. 

In fact if you build unbaptized binary, it will panic due to add overflow. But I am not quite sure if the range (42.0..42.0) is correct.

Back to Plotters bug, the fix is quite straightforward: we should handle the zero sized range differently, since there's no better way for us to figure out the best set of tick marks on the axis besides return that single value directly. 

I am going to fix it on Plotters side and publish a new version ASAP. 

for @mb 's issue, I guess it's one of the data point is NaN. Not sure why this is the case.

I will let you guys know once I fixed the issue on my side.

**Comment by lemmih** [comment]
Fixed in 0.3.5. Thank you @elagergren-spideroak and everyone involved.
